### PR TITLE
[let_chains, 1/6] Remove hir::ExprKind::If

### PR DESCRIPTION
**Pull request by Centril**
Per https://github.com/rust-lang/rust/issues/53667#issuecomment-471583239.

r? @oli-obk

### CODE REVIEW

**Comment by petrochenkov** [comment]
IIRC, I got some lifetime issues after trying to desugar *all* `if`s into `match`es (rather than those containing `is`), with `RefCell`s being held for too long (or too little?) and panicking etc.
Don't remember the details right now, perhaps it's fixable.

**Comment by petrochenkov** [comment]
This seems like a nice simplification even regardless of if-let chains, we only need to make sure that it doesn't regress 1) codegen 2) compile times, and 3) diagnostics.

Similarly, desugaring `while` into `loop` is also required for implementing pattern matching expressions, but is also a nice simplification otherwise.

(I'll try to review the code later today.)

**Comment by petrochenkov** [comment]
On the other hand, the desugaring for pattern matching expressions at the AST/HIR border turned out incredibly hacky, when I tried it.
It also made impossible to support those expressions in pattern guards.
```rust
match thing {
    PAT(y) if let Some(x) = y {
        print(x); // Nope
    }
}
```
So it's a pretty real possibility that it't better done somewhere at HIR->MIR time.

**Comment by Centril** [comment]
> This seems like a nice simplification even regardless of if-let chains, we only need to make sure that it doesn't regress 1) codegen 2) compile times, and 3) diagnostics.

I think I fixed all diagnostics issues that I had... some even got improved. :) As for codegen there are issues in https://github.com/rust-lang/rust/pull/59288#issuecomment-474230627 wrt. mir-opt tests (thus the WIP) that I don't really understand yet.

> Similarly, desugaring `while` into `loop` is also required for implementing pattern matching expressions, but is also a nice simplification otherwise.

Yup, I'd like to do that as step 4 in the process after having done the `if` part of the RFC fully. Maybe even remove `continue` from HIR after step 6...

> (I'll try to review the code later today.)

:)

> On the other hand, the desugaring for pattern matching expressions at the AST/HIR border turned out incredibly hacky, when I tried it.

When I tried it out with LBV + a fold in lowering it didn't feel so bad. Haven't thought about the implications of if-let-guards + if-let-chains tho wrt. lowering. :)

**Comment by Centril** [comment]
(I believe @oli-obk is looking into the mir-opt stuff)

File: src/librustc/hir/mod.rs
Comment (oli-obk) [inline src/librustc/hir/mod.rs]: Can this `Precedence` variant be removed now or is it needed for the AST?

File: src/librustc_typeck/check/_match.rs
Comment (oli-obk) [inline src/librustc_typeck/check/_match.rs]:
```suggestion
                        _ => self.cause(expr.span, MatchExpressionArm {
```

File: src/librustc_typeck/check/demand.rs
Comment (oli-obk) [inline src/librustc_typeck/check/demand.rs]:
```suggestion
            return expected == self.tcx.types.bool;
```

File: src/librustc_typeck/check/mod.rs
Comment (oli-obk) [inline src/librustc_typeck/check/mod.rs]: leave this on separate lines as it was before

File: src/librustc/hir/mod.rs
Comment (Centril) [reply src/librustc/hir/mod.rs]: I believe AST has its own https://doc.rust-lang.org/nightly/nightly-rustc/syntax/ast/struct.Expr.html#method.precedence so it shouldn't be needed.

File: src/librustc_typeck/check/demand.rs
Comment (Centril) [reply src/librustc_typeck/check/demand.rs]: 😅 Thanks!

**Comment by oli-obk** [comment]
@rust-timer build

**Comment by oli-obk** [comment]
@rust-timer build bf82f43c83ada4fffaf93a633dfce5c7f55d57fe

**Comment by Centril** [comment]
@rust-timer build aef7b6f

**Comment by Centril** [comment]
@rust-timer build aef7b6fff148779461e6b8403c70e3f1c47f33dd

**Comment by Centril** [comment]
@rust-timer build 6a4b8678ecc14fe00b78b06eb7e9f272680bba37

**Comment by Centril** [comment]
@rust-timer build 2e9b842658ee1c39ab844c5a2ae57732ca9d1a82

**Comment by Centril** [comment]
Skipping match checking for if-exprs doesn't seem to have had any notable effect on perf.

**Comment by Centril** [comment]
@oli-obk This is now ready for review... :)

File: src/librustc_typeck/check/_match.rs
Comment (oli-obk) [inline src/librustc_typeck/check/_match.rs]:
```suggestion
    /// In the following example, `discrim_span` corresponds to the `a + b` expression:
```

File: src/test/incremental/hashes/if_expressions.rs
Comment (oli-obk) [inline src/test/incremental/hashes/if_expressions.rs]: What causes this regression? I don't think it should happen. Is it the `_` pattern requiring a type?

File: src/test/incremental/hashes/if_expressions.rs
Comment (oli-obk) [inline src/test/incremental/hashes/if_expressions.rs]: nope, can't be the `_` pattern, because that exists here, too. What changed for `if let`?

**Review by oli-obk — CHANGES_REQUESTED** [review]
Can you have a look if https://github.com/rust-lang/measureme/tree/master/summarize differs between before/after your PR, maybe we can track down the 1-2% regressions this way

File: src/test/incremental/hashes/if_expressions.rs
Comment (oli-obk) [reply src/test/incremental/hashes/if_expressions.rs]: Don't mind me... I misread what's going on.

**Comment by oli-obk** [comment]
Nominating for discussion at the compiler mtg. This causes up to 3% perf regressions. You'd see the same regression if user code moved from `if foo { bar } else { boo }` to `match foo { true => bar, false => boo, }`. I do believe both should compile equally fast (they do result in the same MIR after our trivial cleanup optimizations after all).

**Comment by Centril** [comment]
@oli-obk Rebased :)

(+ did a small improvement with the addition of `fn stmt(...)` in `lowering.rs`)

File: src/librustc_typeck/check/_match.rs
Comment (eddyb) [inline src/librustc_typeck/check/_match.rs]: Do not use desugaring information to affect the semantics, *only* adjust diagnostics.
This should be handled by generating a type-ascription `cond: bool` expression, for `if cond {...}`.

**Review by eddyb — CHANGES_REQUESTED** [review]

File: src/librustc_typeck/check/_match.rs
Comment (Centril) [reply src/librustc_typeck/check/_match.rs]: (As discussed on Discord in private conversation...)

So I tried both `DropTemps($cond): bool` and `DropTemps($cond: bool)` and both solutions has the problem that `$expr: $type` does not take coercions into account which means that `loop { if break () {} }` which compiles today stops compiling because it requires a coercion `! -> bool`. [As we know](https://github.com/rust-lang/rust/issues/23416), type ascription is sadly broken (it should allow that coercion to happen but it requires an exact match). Moreover, both solutions will likely require more hacks to improve diagnostics again which regressed.

I see three options to solve this in a different manner:
1. Use `$cond as bool` -- this seems more complex than the hack.
2. Embed `Option<PrimTy>` into `DropTemps` and handle the coercion logic in here; this is probably still more complicated than the hack.
3. Remove the hack and accept the semantic change. (requires T-Lang sign-off + FCP)

As it stands right now, the hack is fairly simple and isn't that hard to remove later. The hack is also well tested in the PR so we can be sure there are no correctness issues. However, it does make `if` into something other than a pure desugaring. That's inelegant, but in my view it works as a stop-gap.

I would like to consider option 3. eventually because it improves usability but this is in my view out of scope for this PR.

cc @nikomatsakis -- @eddyb wanted you to sign off on the hack.

Comment (oli-obk) [reply src/librustc_typeck/check/_match.rs]: > Do not use desugaring information to affect the semantics, only adjust diagnostics.

I thought only `Span` expansion info should not be used to affect semantics. If `MatchSource` should only be used for diagnostics, would it make sense to move all of that info into `ExpansionInfo`?

I also like option 3, is there already an issue about that?

File: src/librustc_typeck/check/mod.rs
Comment (oli-obk) [inline src/librustc_typeck/check/mod.rs]: we may need to turn this match into a loop some day, right now we have an invariant though that `DropTemps` can't contain another `DropTemps` directly.

**Review by oli-obk — APPROVED** [review]
r? @nikomatsakis for HACK signoff

File: src/librustc_typeck/check/_match.rs
Comment (Centril) [reply src/librustc_typeck/check/_match.rs]: @oli-obk There's no issue about that to my knowledge as there wasn't a implementation-driven need before. But now that `if` is almost a pure desugaring to `match` there's more justification for doing it (aside from the usability benefits).

One drawback to 3. is that we no longer have `ExpectHasType(bool)` to trigger good diagnostics for `if x = y {`. If we care deeply about that diagnostic trigger we might need to generate a skeleton from the pattern so that we get the expectation back.

For now however, I think we shouldn't try to do too much in this PR. I've left a `FIXME` addressed at rust-lang/lang-team to think about the issue later.

Comment (nikomatsakis) [reply src/librustc_typeck/check/_match.rs]: In what way is using `$cond as bool` more complex than the hack?

Comment (Centril) [reply src/librustc_typeck/check/_match.rs]: @nikomatsakis My thinking is casts have more complex machinery so `$cond as bool` has more logic. The use itself is probably not that complicated but then again the diagnostics need to be tweaked to account for the cast in-between which is probably more complicated than the hack.

Comment (Centril) [reply src/librustc_typeck/check/_match.rs]: To be clear, I'm happy to try out `$cond as bool` if that is desired.

Comment (nikomatsakis) [reply src/librustc_typeck/check/_match.rs]: (Not questioning that it is, it's just not entirely obvious to me)

Comment (Centril) [reply src/librustc_typeck/check/_match.rs]: @nikomatsakis Same! I'm not sure the hack will be best once I actually implement `let_chains`. It may turn out that `$cond as bool` is cleanest when do that but it should be fairly easy to switch if that is the case. =)

**Comment by nikomatsakis** [comment]
@Centril and I talked. I'm ok with "the hack" but I'd like a FIXME issue to be filed where we can discuss possible follow-up. I would be inclined to do one of the following:

- Fix type ascription (I don't think it's all *that* hard to do)
- Or, if that seems too annoying, add a flag in the type ascription HIR that enables the "coercion" behavior we want. That is still a hack, but perhaps less of one. =)

Regardless, I don't think that has to block this PR.

**Comment by Centril** [comment]
Filed FIXME issue, rebased, and did a minor cleanup after the `.await` PR.

@oli-obk Please take a look :)

**Comment by Manishearth** [comment]
Have you looked into how hard it will be to make clippy migrate to these? We have a lot of lints around ifs, and the lowering may lose some context necessary to make these work. Removing lints because they're no longer possible isn't an option.

IMO for such changes where there's a good chance clippy lints will be unfixably broken we should be notified of this _before_ it merges :/

**Comment by Manishearth** [comment]
I appreciate it being split up, though, it's easier to make this evaluation if one thing breaks at a time :)

**Comment by Centril** [comment]
> We have a lot of lints around ifs, and the lowering may lose some context necessary to make these work.

The lowering preserves information about the source; you can check `MatchSource::IfDesugar` to see whether it originated from an `if`.

@oli-obk thought it would work for clippy and as you know, he also works on clippy...

**Comment by Manishearth** [comment]
Ah okay, thanks. I'll see how hard this is to fix.

For the later steps can you work with us to ensure stuff isn't broken
before landing? Mostly just pinging me when the PR is ready to merge so I
can work on a patch

**Comment by Centril** [comment]
> For the later steps can you work with us to ensure stuff isn't broken before landing? Mostly just pinging me when the PR is ready to merge so I can work on a patch

Sure. Fwiw, I expected Oliver to be the "work with us" party... ;)

The next steps will be to introduce `ast::ExprKind::Let` and then to use `'label: { ... break 'label value }`. That will probably be less straightforward but I think we can introduce similar desugaring info as needed. I could introduce `ast::ExprKind::Let` in a separate intermediate step however without connecting the parser (already done in a PR but I could probably split things).

**Comment by Manishearth** [comment]
That's fair, though it's better if more clippy folks know. I'm working on a
fix now.

**Comment by nnethercote** [comment]
This ended up being a non-trivial [perf regression](https://perf.rust-lang.org/compare.html?start=a850a426491e14186af2250549bf41256b5938d2&end=2e9b842658ee1c39ab844c5a2ae57732ca9d1a82).

**Comment by Centril** [comment]
Yep; this was expected. https://github.com/rust-lang/rust/pull/60730 will hopefully reduce some of those regressions.

**Pull request merged**
